src/proc/validator.rs

```diff
@@ -28,8 +81,242 @@ impl Validator {
+            | crate::StorageClass::Uniform => {
+                match var.binding {
+                    Some(crate::Binding::BuiltIn(_)) => {} // validated per entry point
+                    Some(crate::Binding::Resource { group, binding }) => {
+                        if group > MAX_BIND_GROUPS || binding > MAX_BIND_INDICES {
+                            return Err(GlobalVariableError::OutOfRangeBinding);
+                        }
+                    }
+                    Some(crate::Binding::Location(_)) | None => {
+                        return Err(GlobalVariableError::InvalidBinding)
+                    }
+                }
+                if var.interpolation.is_some() {
+                    return Err(GlobalVariableError::InvalidInterpolation);
+                }
+                //TODO: prevent `Uniform` storage class with `STORE` access
+                match types[var.ty].inner {
+                    crate::TypeInner::Struct { .. }
+                    | crate::TypeInner::Image {
+                        class: crate::ImageClass::Storage(_),
+                        ..
+                    } => true,
+                    _ => false,
+                }
+            }
+            crate::StorageClass::Private | crate::StorageClass::WorkGroup => {
+                if var.binding.is_some() {
+                    return Err(GlobalVariableError::InvalidBinding);
+                }
+                if var.interpolation.is_some() {
+                    return Err(GlobalVariableError::InvalidInterpolation);
+                }
+                false
+            }
+        };
+
+        if !is_storage && !var.storage_access.is_empty() {
+            return Err(GlobalVariableError::InvalidStorageAccess);
+        }
+
+        Ok(())
+    }
+
+    fn validate_function(
+        &mut self,
+        fun: &crate::Function,
+        module: &crate::Module,
+    ) -> Result<(), FunctionError> {
+        let resolve_ctx = ResolveContext {
+            constants: &module.constants,
+            global_vars: &module.global_variables,
+            local_vars: &fun.local_variables,
+            functions: &module.functions,
+            parameter_types: &fun.parameter_types,
+        };
+        self.typifier
+            .resolve_all(&fun.expressions, &module.types, &resolve_ctx)?;
+        Ok(())
+    }
+
+    fn validate_entry_point(
+        &mut self,
+        ep: &crate::EntryPoint,
+        stage: crate::ShaderStage,
+        module: &crate::Module,
+    ) -> Result<(), EntryPointError> {
+        if ep.early_depth_test.is_some() && stage != crate::ShaderStage::Fragment {
+            return Err(EntryPointError::UnexpectedEarlyDepthTest);
+        }
+        if stage == crate::ShaderStage::Compute {
+            if ep
+                .workgroup_size
+                .iter()
+                .any(|&s| s == 0 || s > MAX_WORKGROUP_SIZE)
+            {
+                return Err(EntryPointError::InvalidWorkgroupSize);
```

Comment:
Shouldn't the error be something like: `OutOfRangeWorkgroupSize`? The word `Invalid` seems a little bit too generic here in my opinion.